Add "respell with sharps", "respell with flats" commands

Reedbeta opened the pull request:
This PR introduces two new commands, called "respell pitches with sharps" and "respell pitches with flats", which adjust the enharmonic spelling of selected notes or the entire score. This is useful to avoid laborious note-by-note editing in cases where, for example, MIDI input has spelled things using sharps but I want it to use flats, or vice versa.

Quick demo video:

https://github.com/user-attachments/assets/92142503-1468-430d-a89f-4958c68c1c05

Along the way, I also fixed a couple of minor issues I noticed with the "optimize enharmonic spellings" command:
- it didn't affect grace notes
- it didn't change all the notes of a tie if they weren't all selected

<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

MarcSabatella commented:
I have nothing against these commands, but I would note that accomplishing this is already quite simple - Up followed by Down respells with flats, and the vice versa for sharps.  Also, if you enter the key signature first, the spellings will follow that - choosing the diatonci spelling where possible, and the closest spelling on the circle of fifths for non-diatonic notes.  This will virtually always do better than blindly spelling everything with flats or everything with sharps.

Reedbeta commented:
> Up followed by Down respells with flats, and the vice versa for sharps.

Ahh, that's a nice trick. I didn't know that. Perhaps it's still worth having the commands since they're more discoverable, though.

> Also, if you enter the key signature first, the spellings will follow that - choosing the diatonci spelling where possible, and the closest spelling on the circle of fifths for non-diatonic notes.

This is true, but not every melody is in just one key, or in a key at all. I was transcribing a line that starts in D major, and then modulates to B flat major, and then various other keys, but without changing the key signature. It would have been useful to have a way to respell with flats for the B flat section.

Reedbeta commented on src/engraving/dom/score.cpp:
```diff
@@ -884,32 +885,72 @@ void Score::spell()
-        for (Segment* s = startSegment; s && s != endSegment; s = s->next()) {
+    std::vector<Note*> notes;
+    for (staff_idx_t i = 0; i < score->nstaves(); ++i) {
+        for (Segment* s = score->firstSegment(SegmentType::All); s; s = s->next1()) {
             track_idx_t strack = i * VOICES;
             track_idx_t etrack = strack + VOICES;
             for (track_idx_t track = strack; track < etrack; ++track) {
                 EngravingItem* e = s->element(track);
                 if (e && e->type() == ElementType::CHORD) {
-                    notes.insert(notes.end(),
-                                 toChord(e)->notes().begin(),
-                                 toChord(e)->notes().end());
+                    Chord* c = toChord(e);
+                    notes.insert(notes.end(), c->notes().begin(), c->notes().end());
+                    for (Chord* g : c->graceNotes()) {
+                        notes.insert(notes.end(), g->notes().begin(), g->notes().end());
+                    }
                 }
             }
         }
-        spellNotelist(notes);
+    }
+    return notes;
+}
```
I added this in order to implement `spellWithSharps` and `spellWithFlats`, but we could also use it to implement `spell`. The only reason I didn't do that is it would result in calling `spellNotelist` on the whole score at once, while the existing implementation does it one staff at a time. That could result in some subtle behavior changes.

Jojo-Schmitz commented:
See also #26130

Reedbeta commented:
> Although if I'm honest, I'd much rather plug effort into making the MIDI keyboard input system more intelligent w.r.to this problem of accidentals. For example, currently when you enter F-Ab-C-Eb, MS Studio writes F-**G#**-C-Eb, which in most cases is not what you'd want.

Yes, making the keyboard input system more intelligent would be nice. I would just point out that the issue is not only about chords. Would this proposed more intelligent system try to detect arpeggiations of common chords as well, for example? Would it "go back" and alter previously entered notes in that case? (Say you have an F and G# in the measure already, and then you go into note input and enter a C and Eb; should it respell the existing G# to an Ab?) What about a melodic line that "feels like" it's in F minor?

It doesn't seem terribly plausible to me that a smart input system could be smart enough to handle all these cases, so I think it will often be necessary for the author to go back and make corrections after inputting notes, even if a smarter input system reduces how often that happens. The commands proposed here are an attempt to make those corrections less laborious to accomplish - as you can apply "respell with flats" or "respell with sharps" to anything from a single chord to a lengthy passage, rather than individually fixing the notes.

> Nevertheless, in the above case, the new commands proposed here do _almost_ solve the problem, which _is_ helpful, although some individual-note correction is still required. It's therefore not a robust solution to the stated problem.

I'm not quite sure what you mean by this? In the examples you gave, selecting the chord and executing "respell with flats" will give the appropriate spelling immediately, AFAICT.

Reedbeta commented:
After having used this in earnest for a couple weeks in my personal build, I think it needs some refinement. Depending on key signature and transposition, it will sometimes use the "wrong" accidental, e.g. sharps when respelling with flats; and sometimes it also comes out with a double sharp or double flat unexpectedly. It also sometimes spells Cb instead of B natural, etc.

Reedbeta commented:
@bkunda I had a thought about a different way to approach this problem. What if the sharp/flat toolbar buttons and associated shortcuts could be used with MIDI input to influence the spelling of entered notes? As far as I can tell, sharp/flat buttons currently do nothing with MIDI input, so this could be a useful extension to reduce the need for fixing up spellings after the fact.

Analogously to how using the computer keyboard, I can press <kbd>+</kbd><kbd>C</kbd> to enter a C#, I could press <kbd>+</kbd> or <kbd>-</kbd> and then press the C#/Db key on the MIDI keyboard, and it would spell it as C# or Db accordingly.

I'd envision these settings being sticky when using MIDI entry (instead of auto-resetting as they do with keyboard entry), and only influencing notes that would require an accidental in the current key signature – so, for example, in a key signature with 2 sharps, C#/Db would always be spelled as C (implicit #). (Perhaps unless C was altered to C natural earlier in that measure? Details like this could be refined in the future.)

If that sounds worth exploring, I can put it into a feature request and take a swing at implementing it.

MarcSabatella commented:
FWIW, I like the idea.  But do note, there is also the "J" command to respell retroactively.  Retroactive respelling is going to be necessary anyhow for imported MIDI, but the proposed behavior of the flat and sharp commands could indeed be useful for step time input, and harmless since as observed they have no effect currently.

Reedbeta commented:
I don't find the J command particularly useful; it respells every note in the selected range, cycling each one through all possible spellings.

Usually when I'm doing this I'm transcribing a whole passage and many pitches are correct already. Selecting and fixing each individual wrong note is too laborious, so I want something I can apply to a whole range, and J just isn't useful for that.

Reedbeta commented:
Maybe a concrete example would be helpful. Here's something I transcribed recently, typical of the kind of music I'm working on:

![image](https://github.com/user-attachments/assets/331e43f5-040c-4ed4-93a1-7580f3834d77)

When I entered this using my MIDI keyboard, initially all the Ab's were spelled G#, all the Db's were C#, etc. As you can see, in the first few measures, across 7 staves, it's all spelled using flats. Then comes a couple of measures where sharps are used; these measures have some D#s which were initally spelled Eb, and A#s which were spelled Bb. Then it's back to flats again.

In this case, I can fix things up afterward very quickly by making a range selection to include the first few measures where flats are used and using 'respell with flats' (or using the up and down arrows to do the same), then selecting the area where sharps are used and using 'respell with sharps', etc.

Or, with the proposed sticky accidental hints for MIDI input, I could use the sticky flat hint for the first few bars of each staff, then change to a sticky sharp, then back to a sticky flat, etc.

Either of those is better for me than finding every pitch that needs respelling and changing them one by one. (This can be accelerated by doing a range selection, then using right click > Select > More... > Same note name, so that I can adjust all the G#'s at once, then all the D#'s, etc. But it's still too laborious in my opinion compared to using one command that respells everything at once.)

For the case of mixing MIDI entry with keyboard/mouse entry: I think it works straightforwardly if MIDI entry doesn't reset the accidental mode after adding a note, while keyboard/mouse entry does. The only thing I can see that might be surprising is if you have an accidental enabled while using MIDI entry and forget to turn it off, and then use the keyboard/mouse (within the same note input session), the accidental might unintentionally apply to the first keyboard/mouse entered note.

Reedbeta commented:
Thanks Casper. I looked everything over and it seems good to me. I squashed a few commits together for a cleaner history, and updated the new submenu title to "Other respelling options".
